packages/babel-parser/src/tokenizer/index.js

```diff
@@ -403,7 +403,44 @@ export default class Tokenizer extends LocationParser {
   // into it.
   //
   // All in the name of speed.
-  //
+
+  // number sign is "#"
+  readToken_numberSign(code: number): void {
+    if (this.state.pos === 0 && this.readToken_interpreter()) {
+      return;
+    }
+
+    const nextPos = this.state.pos + 1;
+    const next = this.input.charCodeAt(nextPos);
+    if (
+      (this.hasPlugin("classPrivateProperties") ||
+        this.hasPlugin("classPrivateMethods")) &&
+      this.state.classLevel > 0
+    ) {
+      ++this.state.pos;
+      this.finishToken(tt.hash);
+      return;
```

Comment:
I think that we can make this function simpler by handling private names and topic references the same way (always throwing if there is a digit).
As it is written now, I think that the nice error thrown in the next `if` won’t be thrown for `#1` references inside classes when the private fields plugin is enabled.

Reply:
The way it's currently set up, the error is thrown elsewhere. I moved that `if` block up before the `classLevel` check, which resulted in this test failure:

> SyntaxError: experimental/class-private-properties/failure-numeric-start-identifier/input.js: Expected error message: Identifier directly after number (2:4). Got error message: Unexpected digit after topic reference: '#2' (2:2)

Itself, not a really big deal, but I realized that putting the `if` upfront doesn't distinguish between class properties & topic tokens. Should we simplify the message so it applies generically to both (e.g. "Unexpected digit after hash token") or fork the logic there based on whether we're in a class, or something else entirely?

Reply:
+1 for `Unexpected digit after hash token`: users only using private properties may not know what a "topic reference" is.

Reply:
Updated the error message.